src/coordinator/server.cc

```diff
@@ -112,6 +115,85 @@ void SerializeNeighbors(SearchIndexPartitionResponse* response,
   }
 }
 
+namespace {
+
+// Context for timer-based retry when waiting for in-flight keys on remote shard
+struct RemoteInFlightRetryContext {
+  SearchIndexPartitionResponse* response;
+  grpc::ServerUnaryReactor* reactor;
+  std::unique_ptr<vmsdk::StopWatch> latency_sample;
+  std::deque<indexes::Neighbor> neighbors;
+  std::unique_ptr<query::SearchParameters> parameters;
+  std::vector<InternedStringPtr> neighbor_keys;
+
+  RemoteInFlightRetryContext(SearchIndexPartitionResponse* resp,
+                             grpc::ServerUnaryReactor* react,
+                             std::unique_ptr<vmsdk::StopWatch> sample,
+                             std::deque<indexes::Neighbor> nbrs,
+                             std::unique_ptr<query::SearchParameters> params,
+                             std::vector<InternedStringPtr> keys)
+      : response(resp),
+        reactor(react),
+        latency_sample(std::move(sample)),
+        neighbors(std::move(nbrs)),
+        parameters(std::move(params)),
+        neighbor_keys(std::move(keys)) {}
+};
+
+void RemoteInFlightRetryTimerCallback(ValkeyModuleCtx* ctx, void* data);
```

Comment:
I you move the definition above `CheckAndHandleRemoteInFlightConflicts` you shouldn't have to declare it here